### ISSUE TITLE
Nuitka might generate unsafe code

### ISSUE THREAD

**Issue by n-bes** [Nuitka/Nuitka#2904]
I've found a few crashes on code generated by Nuitka. Repro steps: https://github.com/n-bes/nuitka-crash-poc

### Example 1

```
# cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.6 LTS (Bionic Beaver)"
```

```
# nuitka3 --version
1.6.5
Commercial: None
Python: 3.6.9 (default, Mar 10 2023, 16:46:00)
Flavor: Debian Python
Executable: /usr/bin/python3
OS: Linux
Arch: aarch64
Distribution: Ubuntu (based on Debian) 18.04.6
Version C compiler: /usr/bin/gcc (gcc).
```

#### Asan
```
==22==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x51d00001e920 at pc 0xaaaab53edf48 bp 0xffffff971280 sp 0xffffff970a70

READ of size 2216 at 0x51d00001e920 thread T0
    #0 0xaaaab53edf44 in __asan_memcpy (/code/cmake-build/crash-nuitka+0x39df44) (BuildId: f53aebe4f7e8c1cf)
    #1 0xaaaab549ad64 in DEEP_COPY_DICT /code/nuitka-build/static_src/HelpersDictionariesGenerated.c:427:17
    #2 0xaaaab57116e0 in main /code/run.c:115:35
    #3 0xffffa453579c in __libc_start_main /build/glibc-L0whdJ/glibc-2.27/csu/../csu/libc-start.c:310
    #4 0xaaaab5350360 in _start (/code/cmake-build/crash-nuitka+0x300360) (BuildId: f53aebe4f7e8c1cf)




0x51d00001e920 is located 0 bytes after 2208-byte region [0x51d00001e080,0x51d00001e920)
allocated by thread T0 here:
    #0 0xaaaab53eff94 in malloc (/code/cmake-build/crash-nuitka+0x39ff94) (BuildId: f53aebe4f7e8c1cf)
    #1 0xffffa49de3cc in _PyObject_Alloc build-shared/../Objects/obmalloc.c:1446:22
    #2 0xffffa49de3cc in _PyObject_Malloc build-shared/../Objects/obmalloc.c:1456:12
    #3 0xffffa49f0a8c in new_keys_object build-shared/../Objects/dictobject.c:537:14
    #4 0xffffa49f0b2c in _PyDict_NewPresized build-shared/../Objects/dictobject.c:1386:16
    #5 0xaaaab5700eb4 in _unpackBlobConstants /code/nuitka-build/static_src/HelpersConstantsBlob.c:566:27
    #6 0xaaaab566bd00 in unpackBlobConstants /code/nuitka-build/static_src/HelpersConstantsBlob.c:1219:5
    #7 0xaaaab566b74c in loadConstantsBlob /code/nuitka-build/static_src/HelpersConstantsBlob.c:1349:5
    #8 0xaaaab5710f58 in main /code/run.c:51:5
    #9 0xffffa453579c in __libc_start_main /build/glibc-L0whdJ/glibc-2.27/csu/../csu/libc-start.c:310
    #10 0xaaaab5350360 in _start (/code/cmake-build/crash-nuitka+0x300360) (BuildId: f53aebe4f7e8c1cf)
```

#### HWAsan
```
==11==ERROR: HWAddressSanitizer: tag-mismatch on address 0xedcdfffedc00 at pc 0xaaaad5f3340c
READ of size 2216 at 0xedcdfffedc00 tags: 17/87 (ptr/mem) in thread T0
Invalid access starting at offset 2208
    #0 0xaaaad5f3340c in __hwasan_memcpy (/code/cmake-build/crash-nuitka+0x27340c) (BuildId: 8ace36754df69c33)
    #1 0xaaaad5fc121c in DEEP_COPY_DICT /code/nuitka-build/static_src/HelpersDictionariesGenerated.c:427:17
    #2 0xaaaad620932c in main /code/run.c:115:35
    #3 0xffffbd31479c in __libc_start_main /build/glibc-L0whdJ/glibc-2.27/csu/../csu/libc-start.c:310
    #4 0xaaaad5f26618 in _start (/code/cmake-build/crash-nuitka+0x266618) (BuildId: 8ace36754df69c33)

[0xedcdfffedc00,0xedcdfffee600) is a small allocated heap chunk; size: 2560 offset: 0

Cause: heap-buffer-overflow
0xedcdfffedc00 is located 0 bytes inside a 2208-byte region [0xedcdfffedc00,0xedcdfffee4a0)
allocated by thread T0 here:
    #0 0xaaaad5f2eb6c in malloc (/code/cmake-build/crash-nuitka+0x26eb6c) (BuildId: 8ace36754df69c33)
    #1 0xffffbd7bd3cc in _PyObject_Alloc build-shared/../Objects/obmalloc.c:1446:22
    #2 0xffffbd7bd3cc in _PyObject_Malloc build-shared/../Objects/obmalloc.c:1456:12
    #3 0xffffbd7cfa8c in new_keys_object build-shared/../Objects/dictobject.c:537:14
    #4 0xffffbd7cfb2c in _PyDict_NewPresized build-shared/../Objects/dictobject.c:1386:16
    #5 0xaaaad61fa2bc in _unpackBlobConstants /code/nuitka-build/static_src/HelpersConstantsBlob.c:566:27
    #6 0xaaaad6176b04 in unpackBlobConstants /code/nuitka-build/static_src/HelpersConstantsBlob.c:1219:5
    #7 0xaaaad61766a0 in loadConstantsBlob /code/nuitka-build/static_src/HelpersConstantsBlob.c:1349:5
    #8 0xaaaad6208e84 in main /code/run.c:51:5
    #9 0xffffbd31479c in __libc_start_main /build/glibc-L0whdJ/glibc-2.27/csu/../csu/libc-start.c:310
    #10 0xaaaad5f26618 in _start (/code/cmake-build/crash-nuitka+0x266618) (BuildId: 8ace36754df69c33)
```

### Example 2

```
OS: Ubuntu 24.04 / CentOS Stream 9
Nuitka: 2.3
```

```
==1==ERROR: HWAddressSanitizer: tag-mismatch on address 0xffff7ddf3bb0 at pc 0xffff7dc06ddc
READ of size 8 at 0xffff7ddf3bb0 tags: f5/00 (ptr/mem) in thread T0
    #0 0xffff7dc06ddc in PyInit_module /code/module.build/module.module.c:1123:26
    #1 0x66def0 in _PyImport_LoadDynamicModuleWithSpec /usr/src/python3.12-3.12.3-1/build-static/../Python/importdl.c:169:9
    #2 0x66d338 in _imp_create_dynamic_impl /usr/src/python3.12-3.12.3-1/build-static/../Python/import.c:3775:11
    #3 0x66d338 in _imp_create_dynamic /usr/src/python3.12-3.12.3-1/build-static/../Python/clinic/import.c.h:506:20
    #4 0x502d4c in cfunction_vectorcall_FASTCALL /usr/src/python3.12-3.12.3-1/build-static/../Objects/methodobject.c:422:24
    #5 0x5652ac in _PyEval_EvalFrameDefault /usr/src/python3.12-3.12.3-1/build-static/Python/bytecodes.c:3254:26
    #6 0x4c2f00 in _PyObject_VectorcallTstate /usr/src/python3.12-3.12.3-1/build-static/../Include/internal/pycore_call.h:92:11
    #7 0x4c2f00 in object_vacall /usr/src/python3.12-3.12.3-1/build-static/../Objects/call.c:850:14
    #8 0x4c4ae4 in PyObject_CallMethodObjArgs /usr/src/python3.12-3.12.3-1/build-static/../Objects/call.c:911:24
    #9 0x58bee8 in import_find_and_load /usr/src/python3.12-3.12.3-1/build-static/../Python/import.c:2779:11
    #10 0x58bee8 in PyImport_ImportModuleLevelObject /usr/src/python3.12-3.12.3-1/build-static/../Python/import.c:2862:15
    #11 0x565f6c in import_name /usr/src/python3.12-3.12.3-1/build-static/../Python/ceval.c:2482:15
    #12 0x565f6c in _PyEval_EvalFrameDefault /usr/src/python3.12-3.12.3-1/build-static/Python/bytecodes.c:2135:19
    #13 0x560070 in _PyEval_EvalFrame /usr/src/python3.12-3.12.3-1/build-static/../Include/internal/pycore_ceval.h:89:16
    #14 0x560070 in _PyEval_Vector /usr/src/python3.12-3.12.3-1/build-static/../Python/ceval.c:1683:12
    #15 0x560070 in PyEval_EvalCode /usr/src/python3.12-3.12.3-1/build-static/../Python/ceval.c:578:21
    #16 0x598f70 in run_eval_code_obj /usr/src/python3.12-3.12.3-1/build-static/../Python/pythonrun.c:1722:9
    #17 0x598f70 in run_mod /usr/src/python3.12-3.12.3-1/build-static/../Python/pythonrun.c:1743:19
    #18 0x67e6f0 in pyrun_file /usr/src/python3.12-3.12.3-1/build-static/../Python/pythonrun.c:1643:15
    #19 0x67e2c4 in _PyRun_SimpleFileObject /usr/src/python3.12-3.12.3-1/build-static/../Python/pythonrun.c:433:13
    #20 0x67e090 in _PyRun_AnyFileObject /usr/src/python3.12-3.12.3-1/build-static/../Python/pythonrun.c:78:15
    #21 0x6890e8 in pymain_run_file_obj /usr/src/python3.12-3.12.3-1/build-static/../Modules/main.c:360:15
    #22 0x6890e8 in pymain_run_file /usr/src/python3.12-3.12.3-1/build-static/../Modules/main.c:379:15
    #23 0x6890e8 in pymain_run_python /usr/src/python3.12-3.12.3-1/build-static/../Modules/main.c:629:21
    #24 0x6890e8 in Py_RunMain /usr/src/python3.12-3.12.3-1/build-static/../Modules/main.c:709:5
    #25 0x688ca4 in Py_BytesMain /usr/src/python3.12-3.12.3-1/build-static/../Modules/main.c:763:12
    #26 0xffff7fb384c0 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #27 0xffff7fb38594 in __libc_start_main csu/../csu/libc-start.c:360:3
    #28 0x5f24ec in _start (/usr/bin/python3.12+0x5f24ec) (BuildId: 18160fe6beb052a7e6830ecc99e313a3498c377d)


Thread: T0 0xeffe00002000 stack: [0xfffff1f19000,0xfffff2719000) sz: 8388608 tls: [0xffff808f1460,0xffff808f2320)

Memory tags around the buggy address (one tag corresponds to 16 bytes):
  0xffff7ddf3300: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf3400: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf3500: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf3600: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf3700: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf3800: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf3900: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf3a00: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
=>0xffff7ddf3b00: 00  00  00  00  00  00  00  00  00  00  00 [00] 00  00  00  00
  0xffff7ddf3c00: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf3d00: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf3e00: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf3f00: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf4000: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf4100: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf4200: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
  0xffff7ddf4300: 00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
Tags for short granules around the buggy address (one tag corresponds to 16 bytes):
  0xffff7ddf3a00: ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..
=>0xffff7ddf3b00: ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  .. [..] ..  ..  ..  ..
  0xffff7ddf3c00: ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..
See https://clang.llvm.org/docs/HardwareAssistedAddressSanitizerDesign.html#short-granules for a description of short granule tags

Registers where the failure occurred (pc 0xffff7dc06ddc):
    x0  0000000000b1a670  x1  0000ffff808f1bb0  x2  0000000000b266d8  x3  0000ffff7de09de0
    x4  0000000000009d90  x5  0000000000b2d8d0  x6  ea00ef3dffffbef0  x7  0000ffff7db92060
    x8  0000000000000000  x9  0000000000000000  x10 0000000000000000  x11 0000000000000000
    x12 9100ed9dfffeba00  x13 0000000000000027  x14 000000000236fa65  x15 fffffffffffff000
    x16 0000ffff7fe3b350  x17 0000000000000007  x18 0000000000000004  x19 0000000000b266b0
    x20 0200efff00000000  x21 f500ffff7ddf3bb0  x22 00000000000000f5  x23 00000ffff7ddf3bb
    x24 0000ffff7f0fa990  x25 000000000070f5b8  x26 0000ffff7dc06ac4  x27 0000ffff7f0fb920
    x28 0000000000b8a278  x29 0000fffff2717e00  x30 0000ffff7dc06de0   sp 0000fffff2717e00
SUMMARY: HWAddressSanitizer: tag-mismatch /code/module.build/module.module.c:1123:26 in PyInit_module
```

**Comment by kayhayen** [comment]
Nuitka 1.6 is not too relevant anymore.

**Comment by n-bes** [comment]
@kayhayen `example 2` based on fresh version: 2.3

**Comment by kayhayen** [comment]
@n-bes You are right, but 1.6 stopped me from reading on.

**Comment by kayhayen** [comment]
At some point a commercial customer provided information about Nuitka and these kinds of reports. I do not recall them being very actual. I think it was copying unused memory and such, resulting in the micro-optimization of the dictionary copy, but it wasn't a real bug in that sense.

I still have what I used to check it, from those days:

```
ASAN_OPTIONS=detect_leaks=0 LDFLAGS="-fsanitize=address" CCFLAGS="-Og -fno-omit-frame-pointer -fsanitize=address -w -DNDEBUG" NUITKA_EXTRA_OPTIONS="--clang --lto=no --unstripped --static-libpython=no" python3.10 tests/basics/run_all.py search
```

I am currently re-running that, with 3.10 and then 3.12 maybe. It seems that HWAsan is something different, can you provide me with the flags to use to get that, I am not all that familiar with any of this.

**Comment by n-bes** [comment]
>  I do not recall them being very actual

False-positive reports are very rare, please, do not ignore them.

> I think it was copying unused memory and such, resulting in the micro-optimization of the dictionary copy, but it wasn't a real bug in that sense.

Coping unused memory is not marked as a problem by ASAN/HWasan, only incorrect memory access (out-of-bounds, use of uninitialized memory, etc.)


> It seems that HWAsan is something different

[Clang docs](https://clang.llvm.org/docs/HardwareAssistedAddressSanitizerDesign.html) describes it.

> Can you provide me with the flags to use to get that, I am not all that familiar with any of this.

All information available in the repo, useful links:
* https://github.com/n-bes/nuitka-crash-poc/blob/init/Dockerfile#L65
* https://github.com/n-bes/nuitka-crash-poc/blob/init/Dockerfile#L90
* https://github.com/n-bes/nuitka-crash-poc/blob/init/docker-compose.yaml#L136

**Comment by n-bes** [comment]
@kayhayen reopen?

**Comment by kayhayen** [comment]
I actually resolved them all, what I was saying is that they were not observable issues, like allocating or copying more memory than necessary, but never in a way that causes overflow.

**Comment by kayhayen** [comment]
I am not sure, if I have `AArch64` hardware, does my M1 mac quality for that? Otherwise I have 64 bit ARM, but not too much. The code you compiled looks very small-ish. Can you maybe provide the lines around the one where your stack trace points to, if I am never going to have it?

Also, I only ever checked programs like that, never modules, I might have to do that too.

**Comment by kayhayen** [comment]
From what I am reading, ASAN should be good enough and not cost more than HWASAN, just at a lower cost, so I don't have to worry about using it. Should issues be apparent in either? Can you confirm to me, that my theory holds for the 2.3 issue you are seeing?

**Comment by kayhayen** [comment]
It seems that I was getting linker errors for module builds. I updated from llvm-15 to llvm-19, and now it no longer works at all for me locally on my Ubuntu 22.04. So I downgraded to llvm-14 as in Ubuntu, and now it works again. The LLVM snapshots appears to be only so good.

However, it appears I need to build a Python that is built with ASAN in mind. I cannot quite see that in your Dockerfile, though, and maybe I need to try hwsanitize, maybe it doesn't need it.

**Comment by kayhayen** [comment]
So, I think I convinced it to compile semi-properly for packages

```
LD_LIBRARY_PATH=/usr/lib/llvm-14/lib/clang/14.0.0/lib/linux/ ASAN_OPTIONS=detect_leaks=0 LDFLAGS="-fsanitize=hwaddress -g -shared-libasan -fuse-ld=lld -Wl,--no-relax" CCFLAGS="-Og -fno-omit-frame-pointer -fsanitize=hwaddress -w -DNDEBUG" NUITKA_EXTRA_OPTIONS="--clang --lto=yes --unstripped" python3.10 tests/packages/run_all.py resume
```

However, getting `FATAL: HWAddressSanitizer requires a kernel with tagged address ABI.` in my WSL. Guess I need to spin up some Ubuntu it seems.

**Comment by n-bes** [comment]
> I actually resolved them all, what I was saying is that they were not observable issues, like allocating or copying more memory than necessary, but never in a way that causes overflow.

I found a lot of null dereferences on generated code by static analysis tool and it is the main reason why I just run it with ASAN/HWasan and prove it.

> I actually resolved them all, what I was saying is that they were not observable issues, like allocating or copying more memory than necessary, but never in a way that causes overflow.

I'll try to reproduce `example 1` on latest nuitka version.

> I am not sure, if I have AArch64 hardware, does my M1 mac quality for that? Otherwise I have 64 bit ARM, but not too much.

I don't know about m1, my local setup is M3 Max with Docker Desktop (kernel version is `Linux d3e7e642290b 6.6.26-linuxkit`)

> It seems that I was getting linker errors for module builds. I updated from llvm-15 to llvm-19, and now it no longer works at all for me locally on my Ubuntu 22.04. So I downgraded to llvm-14 as in Ubuntu, and now it works again. The LLVM snapshots appears to be only so good.

I prepared Ubuntu 22.04 with clang-19, and it works (and crashes): https://github.com/n-bes/nuitka-crash-poc/blob/init/Dockerfile#L65


> Can you maybe provide the lines around the one where your stack trace points to, if I am never going to have it?
> Also, I only ever checked programs like that, never modules, I might have to do that too.

See:
source code: https://github.com/n-bes/nuitka-crash-poc/tree/init/ubuntu-2204-hwasan-env-with-nuitka/module.build
stack trace: https://github.com/n-bes/nuitka-crash-poc?tab=readme-ov-file#ubuntu-2204-hwasan-env-with-nuitka

> However, it appears I need to build a Python that is built with ASAN in mind. I cannot quite see that in your Dockerfile, though, and maybe I need to try hwsanitize, maybe it doesn't need it.

I think that compiled python interpreter from sources with ASAN should be enough to detect problems. It is my first HWAsan launch, so ... ASan also replace allocator and can't detect the problem, for now we may only guess about results.

**Comment by kayhayen** [comment]
The code in question is this:

```
    if (_Py_PackageContext != NULL) {
        module_full_name = _Py_PackageContext;
    }
```

For 3.12 that is using this:

```
#if PYTHON_VERSION >= 0x3c0
#define _Py_PackageContext (_PyRuntime.imports.pkgcontext)
#endif
```

and 

```
static char const *module_full_name = %(module_name_cstr)s;
```

I am not getting what that should mean in terms of a illegal write. It would appear that in libpython it's defined similar,

```const char * pkgcontext;```

So, can you explain what this means. Is it complaining that I am copying a pointer into a DLL (the extension module) that is exported from the main process?

**Comment by kayhayen** [comment]
I didn't see the use of `llvm.sh`. Unfortunately my Virtualbox hates even after an update the Ubuntu 24.04 installation so hard, it hurts, kind of surprising. I avoid working with Docker when I can, I have had issues on macOS with architecture mismatches, but you probably are using arm there.

**Comment by kayhayen** [comment]
> I found a lot of null dereferences on generated code by static analysis tool and it is the main reason why I just run it with ASAN/HWasan and prove it.

These would probably be of interest, but lets just say, the stacktraces you provided, are clearly not one of those. It does a NULL check after all, but even if it didn't that would copy a NULL pointer and that would be an issue only later, most definitely a crash with segfault.

**Comment by kayhayen** [comment]
I gave macOS a try, and for packages, it refuses to work, because the Python is not instrumented, but for programs, it seems to not find any stuff with `-fsanitize=address` and that's good and all. The hwaddress is not support for arm macOS proper it seems.

So, I think you reproduced the results I once had with 1.6 and I don't feel like their is new evidence. The report given I would have to read up what it could even mean, but I am like 100% sure that for pre-3.12 it is definitely not a bug, and for 3.12 it is working pretty well for getting the right name to use. Also, in your example, it probably should be NULL, I can check that.

Something that might be wrong, now that I think of it, that it might have to do a `strdup` if it was to use it that much later, but I think it doesn't do that, i.e. when the module returns, the usage is stopped or should be. But I don't think that is what the message means, what does it mean?

**Comment by kayhayen** [comment]
I have pushed to the factory branch a change, where the assignment only happens when the names differ and now it is done via `strcpy` so your stacktrace should change. I am still not sure how valid it is, but checking `PyModule_Create` doesn't see to remember the string other than as Python objects, but the `md_def` remains in the module, but it seems the name is never used, but it's better to be safe than sorry, dangling pointers are not a good thing.

Please retest with current factory state, and lets see what pops up. I would expect nothing, since I didn't understand hwaddress to be better. Yet, I didn't have a chance to see module mode myself. But I am not entirely inclined right now to reproduce what I feel is a false alarm. Also, does it only give a single report at a time?

**Comment by n-bes** [comment]
> python3.10

In Ubuntu 22.04 python version is 3.10

> I'll try to reproduce example 1 on latest nuitka version.

Do not reproduced with gcc (ASAN) and factory version of nutika. I'll try with clang and HWASAN.

> So, can you explain what this means. Is it complaining that I am copying a pointer into a DLL (the extension module) that is exported from the main process?

I think that something happens while trying to load module in memory, not related with copying of pointer.


> Please retest with current factory state, and lets see what pops up.

Ready: 
* https://github.com/n-bes/nuitka-crash-poc/tree/init?tab=readme-ov-file#ubuntu-2404-hwasan-factory-env-with-nuitka
* https://github.com/n-bes/nuitka-crash-poc/blob/init/docker-compose.yaml#L187

> Also, does it only give a single report at a time?

Yes, it is stopped on first error.

**Comment by kayhayen** [comment]
I am not convinced that your report is not an issue with hwsanitizer at all. And I would ask you to use not module mode, but executable mode and find something. It seems I wasn't able to do that, and it seems while it links an extension module, that doesn't mean, it likes being run inside the DLL only while talking to the outside. So until there is new evidence that doesn't just look like a false alarm, I am closing this, sorry.

**Comment by kayhayen** [comment]
I would of course be interested in your static analysis that claims there might be errors in the code, even if providing the evidence for it is not as easy as you hope to be, it could be something and I can of course decide if it's true.

**Comment by n-bes** [comment]
@kayhayen I added results for to the repository in two formats:
- [plain-text](https://github.com/n-bes/nuitka-crash-poc/blob/init/ubuntu-2404-env-with-nuitka/svace/nuitka-poc-svace.txt)
- [sarif](https://github.com/n-bes/nuitka-crash-poc/blob/init/ubuntu-2404-env-with-nuitka/svace/nuitka-poc-svace.svres.sarif.json)

**Comment by kayhayen** [comment]
So, the ones about unreachable code in generated code make no sense:

```
* UNREACHABLE_CODE: The condition in ternary operator is always true at HelpersComparisonGe.c:364.
    ternary operator at /usr/local/lib/python3.12/dist-packages/nuitka/build/static_src/HelpersComparisonGe.c:364
```

This is generated from a Jinja2 template, that is using a bool to nuitka_bool conversion with the explicit intent to let the C compiler resolve this at its compile time rather than creating unnecessarily complex template code. If you can, ignore these things in those files.

Similarily this one makes no sense to me either.

```
    default:
        NUITKA_CANNOT_GET_HERE("invalid PYGEN_ result");
    }
```

I have those to catch implementation errors at runtime. Having a default is a good practice, is it not.


Then those

```
* TAINTED_PTR.MIGHT: Variable 'home_path' is obtained from an untrusted source at HelpersFilesystemPaths.c:951 by calling function 'getenv'. The variable is used in a vulnerable operation at HelpersSafeStrings.c:57 by passing as 2nd parameter to function 'appendStringSafe' at HelpersFilesystemPaths.c:963. Please double check if input string is properly validated in error-prone manner
    [vulnerable operation] Variable 'home_path' is passed to function 'appendStringSafe' as 2nd parameter at /usr/local/lib/python3.12/dist-packages/nuitka/build/static_src/HelpersFilesystemPaths.c:963
    [tainted pointer] Call of 'getenv' at /usr/local/lib/python3.12/dist-packages/nuitka/build/static_src/HelpersFilesystemPaths.c:951
```

These functions e.g. `appendStringSafe` will abort on overflows, and thus the usage is safe.

And this

```
            coercion c2 =
                (type2->tp_as_number != NULL && NEW_STYLE_NUMBER_TYPE(type2)) ? type2->tp_as_number->nb_coerce : NULL;
```

Again generated code, `NEW_STYLE_NUMBER` makes it so that this doesn't happen on Python3.12, but it's not the only version this code is used with. Similar story, not making generated code do things, the C compiler will do for us.

This one is actually good:

```
* REDUNDANT_COMPARISON.ALWAYS_FALSE: Redundant comparison '0' > 'res' (0 > {[0, 1]}) is always false at HelpersBuiltin.c:624.
    comparison at /usr/local/lib/python3.12/dist-packages/nuitka/build/static_src/HelpersBuiltin.c:624
```

I probably changed this function to be `bool` result at some point, and code still treats it as `int` result in a few places, I found at least one more spot.

Also this

```
* NO_VA_END: Function 'va_end' was not called before HelpersHeapStorage.c:31 inside function 'Nuitka_PreserveHeap'. (Function 'va_start' was called at HelpersHeapStorage.c:17.)
    [no va_end] no va_end inside function 'Nuitka_PreserveHeap' at /usr/local/lib/python3.12/dist-packages/nuitka/build/static_src/HelpersHeapStorage.c:31
```

Seems not too relevant for the compilers and code I have, but the standard has a "must call" there, so doing it will probably just be something the C compiler knows if it needs it or not.

```
* NEGATIVE_CODE_ERROR.EX: Variable 'wb.len', which might receive a negative value at HelpersOperationInplaceAddUtils.c:71, is used without checking at HelpersOperationInplaceAddUtils.c:91 by calling function 'memcpy'.
    [value use] Variable 'wb.len' is passed to function 'memcpy' as 3rd parameter at /usr/local/lib/python3.12/dist-packages/nuitka/build/static_src/HelpersOperationInplaceAddUtils.c:91
    [error code] Assign the negative value to variable 'wb.len' at /usr/local/lib/python3.12/dist-packages/nuitka/build/static_src/HelpersOperationInplaceAddUtils.c:71
```

In this specific case, the function is known to always work, and in fact `wb.len` may not make sense to be set. Adding an `assert(wb.len >= 0);` can of course affirm that.

This one is in line with what Python does:

```
* HANDLE_LEAK.EX: The handle 'handle' is created at MetaPathBasedLoader.c:766 by calling function 'dlopen' and lost at MetaPathBasedLoader.c:823.
    [acquire] Call of 'dlopen' at /usr/local/lib/python3.12/dist-packages/nuitka/build/static_src/MetaPathBasedLoader.c:766
    [leaked] leak at /usr/local/lib/python3.12/dist-packages/nuitka/build/static_src/MetaPathBasedLoader.c:823
```

Since we use the code from the DLL with `dlsym`, releasing it seems not possible.

This one

```
* BUFFER_SIZE_MISMATCH: Pointer 'args', referencing a memory location of size 8 bytes, is passed as a parameter to function 'memcpy' at HelpersCallingGenerated.c:1577 by calling function 'memcpy' with a size parameter that is always equal to 16. This can lead to a buffer overflow.
    Variable 'args' is passed to function 'memcpy' as 2nd parameter at /usr/local/lib/python3.12/dist-packages/nuitka/build/static_src/HelpersCallingGenerated.c:1577
```

Seems to be about a false alarm, the memcpy is done on a `    PyObject *const *args = &PyTuple_GET_ITEM(pos_args, 0);
` and two array elements are intended to be copied here ```memcpy(python_pars, args, 2 * sizeof(PyObject *)); ``` which would be fine. I am not sure if helping the tool via `&args[0]` can help it, this is just a wrong detection. The size is statically known to be 2 and even asserted in this generated code `assert(PyTuple_GET_SIZE(pos_args) == 2);` so that is false alarm.

There is one on `Nuitka_Generator_send` which checks for NULL, but only once. That is indeed unnecessary, because it's used as a single argument function, so the one check there is can actually be removed and replaced by an assertion I guess.

**Comment by kayhayen** [comment]
I have a bunch of changes on factory, I invite you to re-run this once it hits develop or now. Factory is unusually a bit unstable to run, but I guess your analysis need not care. If possible ignore list the generated C helper issues, that I addressed. For the `LOOKUP_ATTRIBUTE`, I mean to add an asserting variation, as for some values, it's known they will not raise an error. For the `UNICODE_APPEND` one I need to find a strategy, but it seemed the lines didn't quite match, and it's not important yet.

**Comment by kayhayen** [comment]
The example you gave is a memcpy with size 0 on a NULL pointer, and I don't want to make it slower by adding a check. However, since it's common not to have a closure, I consider specializing this case, as I have done before, or just outright attaching the closure to the function after it has been created. The `memcpy(NULL, NULL, 0)` is not defined behavior according to C standards, but I won't care much about that. I am adding an assertion that actually confirms that like `assert(closure_given == 0 || closure != NULL);` and maybe it's going to help.

**Comment by kayhayen** [comment]
I don't think I looked at HWSan results with findings yet. The module mode is not suited. Have you produced results that I have not discovered yet. I only looked at the static analysis results recently.

This static analysis tool is not reliable and doesn't know enough about the code. I am done with looking at it. What remains are false alarms to me. The `LOOKUP_ATTRIBUTE` stuff, I have taken note of to be dealt with when I write Jinja2 template code to generate the existing code helpers, such that a expected-not-possible-to-fail one will be easy to add, but that's not now. It is however not a possible issue to not check for NULL there.

Part of the changes are in the hotfix, part are in develop, and a few minor things were added for factory just now.
